### PR TITLE
feat: delete product

### PR DESCRIPTION
**Pull request by davidolanrewaju**
## What does the PR do?
This PR addresses issue #77 by implementing the functionality to delete a single product from the database.

#### To test this feature, you can send a DELETE request to the following endpoint:

``` DELETE /api/v1/products/{id} ``` - where {id} is the ID of the product you wish to delete.

### Success Response 
If the product is deleted successfully, you will receive the following response structure:
```
{
  "status_code": 200,
  "message": "Product deleted successfully",
}
```

### "Product Does Not Exist" Response
If the product doesn't exist in the database, the following response will be received:
```
{
	error: 'Product not found',
        message: 'The product with ID ${id} does not exist',
}
```

### Invalid Authentication Response
If the request is made by an unauthenticated or unauthorized user, the following response will be received:
```
{
	"error": "Invalid request",
	"message": "User not authenticated or authorized"
}
```

## Task
- [x] Implement authentication logic to ensure the client is signed in
- [x] Grant user the ability to delete registered products
- [x] Ensure the endpoint correctly identifies if the client is signed in
- [x] Test the endpoint correctly deletes any single product selected by client.

## Referenced Issue
[[FEAT] : Delete product API - Backend](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/77)

### CODE REVIEW

File: src/modules/organisations/tests/mocks/organisation.mock.ts [hunk 1]
```diff
@@ -23,6 +19,7 @@ export const createMockOrganisation = (): Organisation => {
     owned_organisations: [],
     created_organisations: [],
     user_type: UserType.ADMIN,
+    products: [] as Product[], // Add the products property
```
Comment (Cobidex) [inline src/modules/organisations/tests/mocks/organisation.mock.ts:22]: Dangling comment, best removed

File: src/modules/product-category/entities/product-category.entity.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { Product } from '../../products/entities/product.entity';
+import { AbstractBaseEntity } from '../../../entities/base.entity';
+import { ApiProperty } from '@nestjs/swagger';
+import { ManyToOne } from 'typeorm';
+
+import { Column, Entity } from 'typeorm';
+
+@Entity()
+export class ProductCategory extends AbstractBaseEntity {
+  @ApiProperty()
```
Comment (Cobidex) [inline src/modules/product-category/entities/product-category.entity.ts:10]: Make documentation more descriptive.

File: src/modules/product-category/product-category.controller.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  findOne(@Param('id') id: string) {
+    return this.productCategoryService.findOne(+id);
+  }
+
+  @Patch(':id')
+  update(@Param('id') id: string, @Body() updateProductCategoryDto: UpdateProductCategoryDto) {
+    return this.productCategoryService.update(+id, updateProductCategoryDto);
+  }
+
+  @Delete(':id')
```
Comment (Cobidex) [inline src/modules/product-category/product-category.controller.ts:30]: Add docs

File: src/modules/product-category/product-category.controller.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import { Controller, Get, Post, Body, Patch, Param, Delete } from '@nestjs/common';
+import { ProductCategoryService } from './product-category.service';
+import { CreateProductCategoryDto } from './dto/create-product-category.dto';
+import { UpdateProductCategoryDto } from './dto/update-product-category.dto';
+
+@Controller('product-category')
```
Comment (Cobidex) [inline src/modules/product-category/product-category.controller.ts:6]: Add docs

File: src/modules/product-category/entities/product-category.entity.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import { Product } from '../../products/entities/product.entity';
+import { AbstractBaseEntity } from '../../../entities/base.entity';
+import { ApiProperty } from '@nestjs/swagger';
+import { ManyToOne } from 'typeorm';
+
```
Comment (Homoakin619) [inline src/modules/product-category/entities/product-category.entity.ts:5]: extra spaces

File: src/modules/products/entities/product.entity.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import { AbstractBaseEntity } from '../../../entities/base.entity';
+import { User } from '../../user/entities/user.entity';
+import { ProductCategory } from '../../product-category/entities/product-category.entity';
+import { ApiProperty } from '@nestjs/swagger';
+
```
Comment (Homoakin619) [inline src/modules/products/entities/product.entity.ts:5]: extra space

File: src/modules/products/products.controller.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import { Controller, Post, Body, Delete, Param, UseGuards } from '@nestjs/common';
+import { ProductsService } from './products.service';
+import { CreateProductDto } from './dto/create-product.dto';
+import { ApiTags, ApiBody, ApiParam, ApiOperation, ApiResponse, ApiBearerAuth } from '@nestjs/swagger';
+import { AuthGuard } from '../../guards/auth.guard';
+
+@UseGuards(AuthGuard)
```
Comment (Homoakin619) [inline src/modules/products/products.controller.ts:7]: remove this, the auth guard is enabled globally by default

File: src/modules/products/products.service.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+import { Product } from './entities/product.entity';
+
+@Injectable()
+export class ProductsService {
+  constructor(
+    @InjectRepository(Product)
+    private productRepository: Repository<Product>
+  ) {}
+
+  async remove(id: string) {
```
Comment (Cobidex) [inline src/modules/products/products.service.ts:14]: Wrap points of database calls in try/catch block with proper error logging in the catch. Reference auth service for example

Comment (davidolanrewaju) [reply src/modules/products/products.service.ts:14]: @Heba-WebDev said using try catch wasn't important

**Review by Homoakin619 — APPROVED** [review]

File: src/modules/products/products.service.ts (hunk 8 above)
Comment (Heba-WebDev) [reply src/modules/products/products.service.ts:13]: You can have a try catch in your controller to log the error or follow @Cobidex, no problem

**Review by Heba-WebDev — APPROVED** [review]

**Pull request closed without merging**